R/learner_randomPlantedForest_classif_rpf.R

```diff
@@ -0,0 +1,130 @@
+#'     This is analogous to `mtry.ratio` in [`classif.ranger`][mlr3learners::mlr_learners_classif.ranger], with
+#'     `max_interaction_limit` adding to `n_features` as an additional constraint.
+#'
+#'
+#' @section Installation:
+#' `remotes::install_github("PlantedML/randomPlantedForest")`
+#'
+#' @templateVar id classif.rpf
+#' @template learner
+#'
+#' @references
+#' `r format_bib("hiabu_2023")`
+#'
+#' @template seealso_learner
+#' @template example
+#' @export
+LearnerClassifRandomPlantedForest = R6Class("LearnerClassifRandomPlantedForest",
+  inherit = LearnerClassif,
+  public = list(
+    #' @description
+    #' Creates a new instance of this [R6][R6::R6Class] class.
+    initialize = function() {
+      param_set = ps(
+        max_interaction = p_int(lower = 0, upper = Inf, default = 1, tags = "train"),
+        max_interaction_ratio = p_dbl(lower = 0, upper = 1, default = 0, tags = "train"),
+        max_interaction_limit = p_int(lower = 1, upper = Inf, default = 30, tags = "train"),
+        ntrees = p_int(lower = 1, upper = Inf, default = 50, tags = "train"),
+        splits = p_int(lower = 1, upper = Inf, default = 30, tags = "train"),
+        split_try = p_int(lower = 1, upper = Inf, default = 10, tags = "train"),
+        t_try = p_dbl(lower = 0, upper = 1, default = 0.4, tags = "train"),
+        loss = p_fct(c("L1", "L2", "logit", "exponential"), default = "exponential", tags = "train"),
+        delta = p_dbl(lower = 0, upper = 1, default = 1, tags = "train"),
+        epsilon = p_dbl(lower = 0, upper = 1, default = 0.1, tags = "train"),
+        deterministic = p_lgl(default = FALSE, tags = "train"),
+        nthreads = p_int(lower = 1, upper = Inf, default = 1, tags = c("train", "threads")),
+        cv = p_lgl(default = FALSE, tags = "train"),
+        purify = p_lgl(default = FALSE, tags = "train")
+      )
+
+      param_set$values = list(loss = "exponential", max_interaction_limit = 30)
+
+      super$initialize(
+        id = "classif.rpf",
+        packages = "randomPlantedForest",
+        feature_types = c("integer", "numeric", "factor", "ordered"),
+        predict_types = c("response", "prob"),
+        param_set = param_set,
+        properties = c("twoclass", "multiclass"),
+        man = "mlr3extralearners::mlr_learners_classif.rpf",
+        label = "Random Planted Forest"
+      )
+    }
+  ),
+  private = list(
+    .train = function(task) {
+      # get parameters for training
+      pars = self$param_set$get_values(tags = "train")
+      # max_interaction_limit is needed but must not be passed to rpf(),
+      # while convert_ratio automatically removes max_interaction_ratio.
+      max_interaction_limit = pars[["max_interaction_limit"]]
+      pars[["max_interaction_limit"]] = NULL
+      n_features = length(task$feature_names)
+
+      pars = convert_ratio(
+        pars, "max_interaction", "max_interaction_ratio",
+        min(n_features, max_interaction_limit)
```

Comment:
`max_interaction_limit` is initially set to 30 so I think it can't be unset, right? And if `max_interaction_ratio` is unset then the `max_interaction` default value of the learner kicks in (which is 1)

Reply:
It can be unset if people insist `learner$param_set$values$max_interaction_ratio = NULL`.

Reply:
Hmmm, so I found:

- If `max_interaction_ratio` is `NULL`, it's ignored and `learner$param_set$values$max_interaction` is used. If that is not declared, the learner uses `max_interaction = 1`.
- If  `max_interaction_limit` is `NULL`, it's ignored and does not cause an issue since `min(n_features, max_interaction_limit)` will then just equal `n_features`, as intended

However, the behavior is ill-defined if both `max_interaction` and `max_interaction_limit` are set _and_ `max_interaction_ratio` is `NULL`:

```r
# Set max_interaction = 3
learner = lrn("classif.rpf", max_interaction = 3)

# Set max_interaction_limit = 2, (intended for use with max_interaction_ratio)
learner$param_set$values$max_interaction_limit = 2

# Unset max_interaction_ratio, so now convert_ratio() does not complain that 
# mutually exclusive max_interaction and max_interaction_ratio are set
learner$param_set$values$max_interaction_ratio = NULL

# Train, check what learner used as params
learner$train(task)
learner$model$params$max_interaction
#> 3
```

So I think `max_interaction_limit` and `max_interaction_ratio` depend on each other to not be `NULL`, so should that be reflected in the `ps` or should that be caught in `.$train()`? 
If the latter, I'm guessing it's safest to `stop()` rather than warn as the user might not know they're running into ill-defined behavior?

I wish this was a straight-forward as for ranger's `mtry`/`mtr.ratio` but unfortunately for this learner it's very useful to limit the degree of interaction, as e.g. `max_interaction = 1000` in a high-dimensional task would take ages to compute and the results would likely not be very useful. Unfortunately I have not come up with a better solution for this yet, and this feels quite hacky.

Reply:
Ideally everything is in the dependencies, but if it is too complicated (and also the case will rarely happen I guess) it is ok to throw an error in `$train()`. 

However I think either the description in the roxygen comments in wrong, or the code is wrong. Lets say we set the max_interaction_ratio to 0.5, there are 4 features and max_interaction_limit = 2. 
In this case, max_interaction should be set to 2. In your code however it returns 0.5 * min(2, 4) = 1

Reply:
Hmm you're right, that's too bad.

The issue is the more I think about it the less I'm sure how to achieve what we need here, i.e. 

a) A way to tune uniformly over `max_interaction` in a benchmark where `n_features` can vary greatly
b) A way to limit `max_interaction` to some upper bound for runtime and just plausibility reasons.

I thought about doing

```r
      # Convert max_interaction_ratio analogously to ranger's mtry.ratio:
      pars = convert_ratio(
        pars, "max_interaction", "max_interaction_ratio",
       n_features
      )
      # Check max_interaction_limit as upper bound, use
      pars$max_interaction = min(pars$max_interaction, max_interaction_limit)
```

But now we still have the issue that a benchmark might tune `max_interaction_ratio` in `[0, 1]`, but if e.g. `max_interaction_limit = 10` and `n_features = 20` then all the tuning budget for ratio in `[0.5, 1]` would be "clumped" as it would all be translated to `max_interaction = 10`.

Hmmm the easiest solution would be to do away with `max_interaction_limit` altogether I guess.

Do you have any better ideas?
If not, I'm tending towards simplifying the learner by removing `max_interaction_limit` and just accepting that tuning/benchmarking will be.. inefficient.

Reply:
Ah, now I understand better where you are coming from. In this case the solution in code was decent and only the explanation in the roxygen comments needs to be corrected, no? (I was just confused that the two did not align). Or what are you unhappy with?

Maybe I would set the inital value of `max_interaction_limits` to `Inf` so users have to opt-in to use it (and hence should read the docs).

Reply:
Ah, that's also an option. I had been thinking about this a lot when I first implemented it and got my head twisted in a knot 😅 I'll rewrite the docs to match the current implementation and your suggestion to set to `Inf`, I guess that'll be fine. 

It's a new/experimental learner, I guess as long as we're transparent about this being still in the tire-kicking phase it's going to be alright like this.